themes/default/content/blog/kubernetes-fundamentals-part-one/index.md

```diff
@@ -0,0 +1,167 @@
+
+Okay, but this all feels very Ops-centric still -- we’re writing YAML, not a programming language, and I promised you a familiar programming language. Let’s use Pulumi to do it with Python instead.
+
+Where before we had a ton of YAML, infrastructure as code tools let us at least cut a window from the Ops silo to the Dev silo, so we can hand each other stuff and talk a bit more and actually work together. You can stand up the infrastructure to deploy your application on Kubernetes right there in the application code! Fortunately, Pulumi has a bunch of different examples for how to do this.
+
+First, you need [Pulumi installed](https://www.pulumi.com/docs/get-started/install/). Then, you need to [configure Kubernetes for Pulumi](https://www.pulumi.com/docs/intro/cloud-providers/kubernetes/setup/). From here, we can deploy to Kubernetes without touching a shred of YAML directly.
+
+Below is an example of the code required to deploy NGINX to Kubernetes, with two replicas, but this time, in more familiar Python instead of YAML.
+
+```python
+import pulumi
+from pulumi_kubernetes.apps.v1 import Deployment, DeploymentSpecArgs
+from pulumi_kubernetes.core.v1 import ContainerArgs, ContainerPortArgs, PodSpecArgs, PodTemplateSpecArgs
+from pulumi_kubernetes.meta.v1 import LabelSelectorArgs, ObjectMetaArgs
+
+config = pulumi.Config()
+nginxLabels = {"app": "nginx"}
+nginxDeployment = Deployment(
+    "nginx-deployment",
+    spec=DeploymentSpecArgs(
+        selector=LabelSelectorArgs(match_labels=nginxLabels),
+        replicas=2 if config.get_int("replicas") is None else config.get_int("replicas"),
+        template=PodTemplateSpecArgs(
+            metadata=ObjectMetaArgs(labels=nginxLabels),
+            spec=PodSpecArgs(
+                containers=[ContainerArgs(
+                    name="nginx",
+                    image="nginx:1.7.9",
+                    ports=[ContainerPortArgs(container_port=80)],
+                )],
+            ),
+        ),
+    ))
+
+pulumi.export("nginx", nginxDeployment.metadata.apply(lambda m: m.name))
+```
+
+In just 25 lines and one file, we’ve set up a deployment with two replicas of a containerized application and exposed it. It’s happening in a language you’re familiar with, and it can live right alongside your application code, rather than keeping everything in silos. If you want to increase the number of replicas, that requires no more effort than it did when you were looking at YAML.
+
+Keeping the Dev team and the Ops team isolated from one another leads to misunderstandings, confusion, and a slower release cycle. At the very least, both parties need to be able to understand each other, and that means Devs learning a bit more about Kubernetes. Pulumi can help you speak Kubernetes in your own language.
+
+This has all been local, though, and you might not be running Kubernetes on-prem. What if you want to abstract things away even further? Tune in next time for Kubernetes on cloud providers!
```

Comment:
This example is all local, though, and you might not be running Kubernetes on-prem. What if you want to abstract things away even further? Tune in next time for Kubernetes on cloud providers!